docs/source/migrate/index.rst

```diff
@@ -0,0 +1,274 @@
+        nonconsuming choice Rollback: ContractId A.Foo
+            with
+                inC : ContractId B.Foo
+                sigs : [Party]
+            controller sigs
+                do
+                    d <- fetch inC
+                    assert $ fromList sigs == fromList (signatory d)
+                    create $ conv d
+
+Currently, we only support one upgrade/rollback contract template pair. In the future we might
+extend the ``migrate`` command with different possible upgrade/rollback templates. The above
+contract template offers a choice to input a contract instance defined in ``foo-1.0.0``, convert it
+with the ``conv`` function and create a new one that follows the data type definition in
+``foo-2.0.0``. The heart of the migration code is the ``conv`` function. It is defined in the DAML
+standard library in the module ``DA.Upgrade`` and has the following type signature:
+
+.. code-block:: daml
+
+  conv : (Generic a repA, Generic b repB, Conv repA repB) => a -> b
+
+From its type signature you can see that it can convert any two data types that are instances of the
+``Generic`` class and whose generic representation can be converted itself.
+
+The ``migrate`` command takes care of defining generic instances for all relevant data types in the
+two packages. In our example, you'll find them in the files ``FooAInstances.daml`` and
+``FooBInstances.daml``.
+
+Generic representations can be converted when they are ``isomorphic``. That means the corresponding
+data types defined in package ``foo-1.0.0`` and ``foo-2.0.0`` have exactly the same shape. For
+example the following two data types are isomorphic:
+
+.. code-block:: daml
+
+  data Either a b = Left a | Right b
+  data UpDown a b = Up a | Down b
+
+while the following data types are not
+
+.. code-block:: daml
+
+  data Either a b = Left a | Right b
+  data Maybe a = Just a | Nothing
+
+When the package ``foo-2.0.0`` contains an extended data type of the ``Foo`` template that is not
+isomorphic, the build will fail. For example, let's assume the ``Foo`` module in the ``foo-2.0.0``
+package has been extended to
+
+.. code-block:: daml
+
+  daml 1.2
+  module Foo where
+
+  template Foo
+    with
+      a : Int
+      b : Text
+      p : Party
+    where
+      signatory p
+
+
+Here is typical error message in this case:
+
+.. code-block:: shell
+
+
+  daml/Foo.daml:22:30: error:
+  • Could not deduce (DA.Upgrade.Conv
+  (DA.Generics.M1
+  DA.Generics.S
+  ('DA.Generics.MetaSel
+  ('DA.Generics.MetaSel0
+  ('Some "p")
+  'DA.Generics.NoSourceUnpackedness
+  'DA.Generics.NoSourceStrictness))
+  (DA.Generics.Rec0 Party))
+  (DA.Generics.S1
+  ('DA.Generics.MetaSel
+  ...
+
+The important hint is that the compiler is not able to deduce that our data type is an instance of
+the ``DA.Upgrade.Conv`` class and hence not convertible. In this case you will have to add your own
+upgrade/rollback templates to ``daml/Foo.daml``, that describe how to convert a contract of the
```

Comment:
daml/Foo.daml` from which package?

Reply:
of the migration project you're building. I'll add it.